contracts/main/bridge/FathomBridge.sol

```diff
@@ -0,0 +1,171 @@
+        _zeroAddressCheck(_bookKeeper);
+        _zeroAddressCheck(_stablecoinAdapter);
+        _zeroAddressCheck(address(_initializerLib));
+        bookKeeper = IBookKeeper(_bookKeeper);
+        stablecoinAdapter = IStablecoinAdapter(_stablecoinAdapter);
+        live = 1;
+        whitelisted[msg.sender] = true;
+        _asterizm_initialize(_initializerLib, true, false);
+    }
+
+    // --- Whitelisting ---
+
+    function addToWhitelist(address _usr) external onlyOwnerOrGov {
+        _zeroAddressCheck(_usr);
+        whitelisted[_usr] = true;
+        emit LogAddToWhitelist(_usr);
+    }
+
+    function removeFromWhitelist(address _usr) external onlyOwnerOrGov {
+        _zeroAddressCheck(_usr);
+        whitelisted[_usr] = false;
+        emit LogRemoveFromWhitelist(_usr);
+    }
+
+    // --- Administration ---
+
+    function setDecentralizedMode(bool _isOn) external onlyOwnerOrGov {
+        isDecentralizedMode = _isOn;
+        emit LogSetDecentralizedMode(_isOn);
+    }
+
+    function setFee(uint256 _newFee) external onlyOwnerOrGov {
+        fixedBridgeFee = _newFee;
+        emit LogSetFee(_newFee);
+    }
+
+    function withdrawFees(address _to) external onlyOwnerOrGov {
+        _zeroAddressCheck(_to);
+        address stablecoin = address(stablecoinAdapter.stablecoin());
+        stablecoin.safeTransfer(_to, stablecoin.balanceOf(address(this)));
+        emit LogWithdrawFees(msg.sender, _to, stablecoin.balanceOf(address(this)));
+    }
+
+    /// Cross-chain transfer
+    /// @notice This function is used to transfer FXD from the source chain to the destination chain.
+    /// It works only when the off chain client module is up and running to listen to this function's event.
+    /// works only when the sender is whitelisted or in decentralized mode
+    /// @param _dstChainId uint64  Destination chain ID
+    /// @param _to address  To address
+    /// @param _amount uint  Amount
+    function crossChainTransfer(uint64 _dstChainId, address _to, uint _amount) external onlyWhitelisted nonReentrant{
+        require(live == 1, "FathomBridge/not-live");
+        require(_amount > fixedBridgeFee, "FathomBridge/amount-less-than-fee");
+        _zeroAddressCheck(_to);
+        address stablecoin = address(stablecoinAdapter.stablecoin());
+        require(stablecoin.balanceOf(msg.sender) >= _amount, "FathomBridge/insufficient-balance");
+
+	    stablecoin.safeTransferFrom(msg.sender, address(this), _amount);
+        
+        stablecoin.safeApprove(address(stablecoinAdapter), 0);
+        uint256 _actualTransferAmount = fixedBridgeFee != 0 ? _amount - fixedBridgeFee : _amount;
+        stablecoin.safeApprove(address(stablecoinAdapter), _actualTransferAmount);
+        stablecoinAdapter.crossChainTransferOut(msg.sender, _actualTransferAmount);
+        stablecoin.safeApprove(address(stablecoinAdapter), 0);
+        // bookkeeping
+        bookKeeper.handleBridgeOut(_dstChainId, _actualTransferAmount);
```

Comment:
Makes sense. Thanks for the wisdom. I would think about this.

Reply:
Your suggestion indeed offers a sequential flow of cross-contract calls, and it’s beautiful. However, I would like to keep the handleBridgeOut/in restricted to the Bridge_role only. This way, any bridge-related functions would be called from the bridge, rather than having FathomBridge be responsible for StablecoinAdapter’s bridge functions and StablecoinAdapter being responsible for bridge functions in BookKeeper.